src/java/org/apache/cassandra/cql3/statements/SelectStatement.java

```diff
@@ -973,10 +974,15 @@ private NavigableSet<Clustering<?>> getRequestedRows(QueryOptions options, Clien
     /**
      * May be used by custom QueryHandler implementations
      */
-    public RowFilter getRowFilter(QueryOptions options) throws InvalidRequestException
+    public RowFilter getRowFilter(QueryOptions options, ClientState state) throws InvalidRequestException
     {
         IndexRegistry indexRegistry = IndexRegistry.obtain(table);
-        return restrictions.getRowFilter(indexRegistry, options);
+        RowFilter filter = restrictions.getRowFilter(indexRegistry, options);
+
+        if (filter.needsReconciliation() && filter.isMutableIntersection() && restrictions.needFiltering(table))
```

Comment:
I tried to order these checks from cheapest to most expensive. In this case, it seemed avoiding the `needFiltering()` call when we don't even have a mutable intersection seemed to make the most sense.